src/analysis/processing/qgsalgorithmspatialjoin.cpp

```diff
@@ -0,0 +1,375 @@
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsalgorithmspatialjoin.h"
+#include "qgsprocessing.h"
+#include "qgsgeometryengine.h"
+#include "qgsvectorlayer.h"
+#include "qgsapplication.h"
+#include "qgsfeature.h"
+#include "qgsfeaturesource.h"
+
+///@cond PRIVATE
+
+
+void QgsSpatialJoinAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ),
+                QObject::tr( "Base Layer" ), QList< int > () << QgsProcessing::QgsProcessing::TypeVectorAnyGeometry ) );
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "JOIN" ),
+                QObject::tr( "Join Layer" ), QList< int > () << QgsProcessing::QgsProcessing::TypeVectorAnyGeometry ) );
+  addParameter( new QgsProcessingParameterEnum( QStringLiteral( "PREDICATE" ), QObject::tr("Geometric predicate"),
+                tags(), true, 0 ) );
+  addParameter( new QgsProcessingParameterField( QStringLiteral( "FIELDS_TO_JOIN" ),
+                QObject::tr( "Fields to join to the base layer (leave empty to copy all fields)" ),
+                QVariant(), QStringLiteral( "FIELDS" ), QgsProcessingParameterField::Any, true, true ) );
+
+  QStringList joinMethods;
+  joinMethods << QObject::tr( "Create separate feature for each matching feature (one-to-many)" )
+          << QObject::tr( "Take attributes of the first matching feature only (one-to-one)" );
+  addParameter( new QgsProcessingParameterEnum( QStringLiteral( "METHOD" ),
+                QObject::tr( "Join type" ),
+                joinMethods, false, 1 ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "DISCARD_NONMATCHING" ),
+                QObject::tr( "Discard records which could not be joined" ),
+                false ) );
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "PREFIX" ),
+                QObject::tr( "Joined field prefix" ), QVariant(), false, true ) );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "JOINED" ), QObject::tr( "Joined layer" ), QgsProcessing::TypeVectorAnyGeometry, QVariant(), true, true ) );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "UNJOINED" ), QObject::tr( "unjoined layer" ), QgsProcessing::TypeVectorAnyGeometry, QVariant(), true, true ) );
+}
+
+QString QgsSpatialJoinAlgorithm::name() const
+{
+  return QStringLiteral( "spatialjoin" );
```

Comment:
```suggestion
  return QStringLiteral( "joinattributesbylocation" );
```

(locked in stable API)